### ISSUE TITLE
test_command_fail_1 fails on debian testing py 3.9 (only)

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#6834]
```
_____________________________ test_command_fail_1 ______________________________

    def test_command_fail_1():
        # Expect that the return code of a failing command is caught,
        # that None is returned as result.
        bc = BatchedCommand(
            cmd=py2cmd(
                """
    print("something")
    exit(3)
                """))
    
        # Send something to start the process
        result = bc("line one")
        assert_equal(bc.return_code, None)
        assert_equal(result, "something")
        result = bc("line two")
>       assert_equal(bc.return_code, 3)

datalad/tests/test_cmd.py:165: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

first = None, second = 3, msg = None

    def assert_equal(first, second, msg=None):
        if msg is None:
>           assert first == second
E           assert None == 3
E             +None
E             -3

datalad/tests/utils_pytest.py:107: AssertionError
```

as was added in 9f066efa6a - @christian-monch any ideas ?

**Comment by yarikoptic** [comment]
on a rebuild of 0.17.1-2 which is functionally the same it went through just fine. Thus adding a label for the best kinds of test failure.

**Comment by christian-monch** [comment]
There is actually an error here. The test is assuming a behavior that does not match the control flow in `BatchedCommand`. A little more detail:`BatchedCommand`will restart the process when `result = bc("line two")` is executed and the process did already exit. This is not anticipated by the test.

**Comment by christian-monch** [comment]
I looked at the history of the restart-code:

The restart code was introduced in PR #334 (https://github.com/datalad/datalad/pull/334), commit 74dada12ce70a5535eb59c926165e4433091436a. The commit was made 2016-01-15 and the PR was merged 2016-01-27

Besides the commit message. no rationale was given:
```
BF/ENH: restart --batched process if it crashes
```

On 2019-04-08 the restart code was factored out into `BatchedCommand`. It seems that the restart code is annex-specific and should probably live in `BatchedAnnex`.

**Comment by yarikoptic** [comment]
Thank you @christian-monch for the investigation! I wish I left more information for that commit. Given the age when that was introduced I think it should indeed be safe to assume that we might now change and just bubble up the error instead of restarting. Use on the wild might point to few left corner cases

**Comment by yarikoptic** [comment]
and this test failed on conda build for 3.10 thus making recent datalad not installable for python 3.10! https://github.com/conda-forge/datalad-feedstock/issues/88 . So I would really welcome to see this issue fixed first by making the test more robust. As @christian-monch identified that it is a race condition.  I am confirming that with

```patch
diff --git a/datalad/tests/test_cmd.py b/datalad/tests/test_cmd.py
index 6096dcd8c..16a9c60e0 100644
--- a/datalad/tests/test_cmd.py
+++ b/datalad/tests/test_cmd.py
@@ -161,6 +161,8 @@ exit(3)
     result = bc("line one")
     assert_equal(bc.return_code, None)
     assert_equal(result, "something")
+    from time import sleep
+    sleep(1)
     result = bc("line two")
     assert_equal(bc.return_code, 3)
     assert_equal(result, None)
```
the issue reproduces quite robustly.  For conda -- I will just patch away this `assert_equal` to 3 to make the build for 3.10.  I am still a little hesitant about stripping away restarting in `maint`, but so far I am leaning that it should be the right thing todo.

**Comment by christian-monch** [comment]
One thing: `BatchedAnnex` and `BatchedCommand`, always automatically start a process with the given command line, if no process exists. That means you can use a single instance of `BatchedCommand` to communicate will different processes consecutively. For example, if a process returns three results and then exits with 0, the process would be silently restarted on the fourth invocation of `BatchedCommand.__call__`, and the same instance now communicates with a different process. I think a lot of code relies on this behavior and it should therefore not be changed. We should only ensure that errors are signaled, i.e. an exception is raised.

**Comment by christian-monch** [comment]
Fixed by PR #7068
